File: utils/permissions.js
// Claw Relay — Permission Scope Management

const SCOPES = {
  read: { label: 'Read', description: 'Snapshots and page info only', risk: 'low' },
  interact: { label: 'Interact', description: 'Click, fill, select elements', risk: 'medium' },
  navigate: { label: 'Navigate', description: 'Change page URLs', risk: 'medium' },
  execute: { label: 'Execute', description: 'Run arbitrary JavaScript', risk: 'high' },
  full: { label: 'Full', description: 'All permissions', risk: 'high' },
};

const DEFAULT_PERMISSIONS = ['read', 'interact'];

function checkPermission(action, grantedPermissions) {
  const permMap = {
    snapshot: 'read',
    status: 'read',
    click: 'interact',
    fill: 'interact',
    select: 'interact',
    navigate: 'navigate',
    evaluate: 'execute',
  };
  const needed = permMap[action] || 'full';
  return grantedPermissions.includes(needed) || grantedPermissions.includes('full');
}

if (typeof module !== 'undefined') {
  module.exports = { SCOPES, DEFAULT_PERMISSIONS, checkPermission };
}

File: relay-core/src/permissions.rs

const SCOPE_MAP: &[(&str, &str)] = &[
    ("snapshot", "read"),
    ("screenshot", "read"),
    ("click", "interact"),
    ("type", "interact"),
    ("fill", "interact"),
    ("press", "interact"),
    ("hover", "interact"),
    ("select", "interact"),
    ("navigate", "navigate"),
    ("close", "navigate"),
    ("evaluate", "execute"),
];

pub fn get_required_scope(action: &str) -> Option<&'static str> {
    SCOPE_MAP.iter().find(|(a, _)| *a == action).map(|(_, s)| *s)
}

pub fn has_permission(scopes: &[String], action: &str) -> bool {
    match get_required_scope(action) {
        Some(required) => scopes.iter().any(|s| s == required),
        None => false,
    }
}

pub const VALID_SCOPES: &[&str] = &["read", "interact", "navigate", "execute"];
